File: python/code_digest.py
#!/usr/bin/env python3
"""
SourceCodeFlatener - Python Implementation
A tool for creating comprehensive digests of source code repositories.
"""

import os
import sys
import argparse
import subprocess
import tempfile
import shutil
import re
from pathlib import Path
from typing import List, Dict, Optional, Set, Tuple
import datetime

# Default settings
DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB
DEFAULT_MAX_OUTPUT_SIZE = 5 * 1024 * 1024  # 5 MB
DEFAULT_EXCLUDE_PATTERNS = [
    "bin/*", "obj/*", "debug/*", "release/*", "node_modules/*", 
    "*.dll", "*.exe", "*.pdb", "*.cache"
]
SUPPORTED_FILE_EXTENSIONS = {
    'ps1', 'cs', 'sln', 'md', 'txt', 'json', 'xml', 'yaml', 'yml', 'py', 'js'
}

class OutputFile:
    """Manages the output file, handling rotation when size limits are reached."""
    
    def __init__(self, output_path: str, max_size: int, project_name: str):
        self.base_path = Path(output_path)
        self.max_size = max_size
        self.project_name = project_name or "Unnamed Project"
        self.part = 1
        
        # Extract base name and extension
        self.base_name = self.base_path.stem
        self.extension = self.base_path.suffix
        self.directory = self.base_path.parent
        
        # Initialize the current output file
        self.current_path = self.base_path
        self._ensure_directory_exists()
        
        # Write initial header
        with open(self.current_path, 'w') as f:
            f.write(f"# Project Digest: {self.project_name}\n")
            f.write(f"Generated on: {datetime.datetime.now()}\n\n")
    
    def _ensure_directory_exists(self):
        """Create the output directory if it doesn't exist."""
        os.makedirs(self.directory, exist_ok=True)
    
    def _check_and_rotate(self):
        """Check if file size exceeds limit and rotate if needed."""
        if not os.path.exists(self.current_path):
            return
            
        if os.path.getsize(self.current_path) > self.max_size:
            self.part += 1
            new_path = self.directory / f"{self.base_name}_part{self.part}{self.extension}"
            self.current_path = new_path
            
            # Write header to new part
            with open(self.current_path, 'w') as f:
                f.write(f"# Project Digest Continued: {self.project_name}\n")
                f.write(f"Generated on: {datetime.datetime.now()}\n\n")
    
    def write_line(self, line: str):
        """Write a line to the output file and rotate if necessary."""
        with open(self.current_path, 'a') as f:
            f.write(line + '\n')
        self._check_and_rotate()
    
    def write_block(self, block: str):
        """Write a block of text to the output file, rotating if necessary."""
        for line in block.split('\n'):
            self.write_line(line)
    
    def get_path(self) -> str:
        """Get the current output file path."""
        return str(self.current_path)


class SourceCodeFlatener:
    """Main class for the SourceCodeFlatener tool."""
    
    def __init__(self, args):
        self.source = args.source
        self.output = args.output
        self.project_name = args.project_name
        self.max_file_size = args.max_file_size
        self.max_output_size = args.max_output_size
        self.include_patterns = args.include_patterns or []
        self.exclude_patterns = args.exclude_patterns or DEFAULT_EXCLUDE_PATTERNS
        self.branch = args.branch
        
        # Statistics
        self.file_count = 0
        self.dir_count = 0
        self.total_bytes = 0
        
        # Temporary repository directory (if cloning)
        self.temp_repo_path = None
        
        # Determine project directory
        if self.source.startswith(('http://', 'https://')):
            self.project_dir = self._clone_repository()
        else:
            if not os.path.isdir(self.source):
                print(f"Error: Directory '{self.source}' does not exist.", file=sys.stderr)
                sys.exit(1)
            self.project_dir = os.path.abspath(self.source)
        
        # Initialize output file
        self.output_file = OutputFile(
            self.output, 
            self.max_output_size, 
            self.project_name
        )
        
        # Update output file with source info
        self.output_file.write_line(f"Source: {self.source}")
        self.output_file.write_line(f"Project Directory: {self.project_dir}")
        self.output_file.write_line("")
    
    def _clone_repository(self) -> str:
        """Clone a git repository to a temporary directory."""
        print(f"Source appears to be a URL: {self.source}", file=sys.stderr)
        self.temp_repo_path = tempfile.mkdtemp(prefix="repo_")
        print(f"Cloning repository {self.source} into {self.temp_repo_path} ...", file=sys.stderr)
        
        cmd = ["git", "clone"]
        if self.branch:
            cmd.extend(["--branch", self.branch])
        else:
            cmd.append("--depth=1")
        
        cmd.extend([self.source, self.temp_repo_path])
        
        try:
            subprocess.run(cmd, check=True, capture_output=True, text=True)
        except subprocess.CalledProcessError as e:
            print(f"Error cloning repository: {e.stderr}", file=sys.stderr)
            if self.temp_repo_path and os.path.exists(self.temp_repo_path):
                shutil.rmtree(self.temp_repo_path)
            sys.exit(1)
        
        return self.temp_repo_path
    
    def _should_ignore(self, rel_path: str) -> bool:
        """Determine if a file or directory should be ignored."""
        # If include patterns are specified and path doesn't match any, ignore it
        if self.include_patterns:
            match_found = False
            for pattern in self.include_patterns:
                if self._glob_match(rel_path, pattern):
                    match_found = True
                    break
            if not match_found:
                return True
        
        # Check if path matches any exclude pattern
        for pattern in self.exclude_patterns:
            if self._glob_match(rel_path, pattern):
                return True
        
        return False
    
    def _glob_match(self, path: str, pattern: str) -> bool:
        """Match a path against a glob pattern."""
        # Convert glob patterns to regex
        regex_pattern = "^" + pattern.replace(".", "\\.").replace("*", ".*") + "$"
        return bool(re.match(regex_pattern, path))
    
    def _get_relative_path(self, path: str) -> str:
        """Get path relative to project directory."""
        return os.path.relpath(path, self.project_dir)
    
    def _process_directory(self, directory: str, indent: str = ""):
        """Process a directory, outputting its structure."""
        rel_path = self._get_relative_path(directory)
        
        if rel_path == ".":
            self.output_file.write_line("[DIR] .")
        else:
            dir_name = os.path.basename(directory)
            self.output_file.write_line(f"{indent}[DIR] {dir_name}")
        
        self.dir_count += 1
        
        # Process directory contents
        for item in os.listdir(directory):
            # Skip hidden files
            if item.startswith('.'):
                continue
            
            full_path = os.path.join(directory, item)
            item_rel = self._get_relative_path(full_path)
            
            # Skip the output file itself
            if os.path.basename(full_path) == os.path.basename(self.output):
                continue
            
            # Skip ignored files/directories
            if self._should_ignore(item_rel):
                continue
            
            if os.path.isdir(full_path):
                self._process_directory(full_path, f"  {indent}")
            else:
                # Check if it's a supported file type
                ext = os.path.splitext(item)[1][1:].lower()
                if ext in SUPPORTED_FILE_EXTENSIONS:
                    self.output_file.write_line(f"{indent}  [FILE] {item}")
    
    def _process_file_contents(self, filepath: str):
        """Process and write the contents of a file."""
        rel_path = self._get_relative_path(filepath)
        ext = os.path.splitext(filepath)[1][1:].lower()
        
        if ext not in SUPPORTED_FILE_EXTENSIONS:
            return
        
        # Skip files that are too large
        if os.path.getsize(filepath) > self.max_file_size:
            self.output_file.write_line("")
            self.output_file.write_line(f"## {rel_path}")
            self.output_file.write_line("[File too large to process]")
            return
        
        self.output_file.write_line("")
        self.output_file.write_line(f"## {rel_path}")
        
        try:
            with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
                content = f.read()
                
                # For markdown files, we might want to strip formatting
                if filepath.endswith('.md'):
                    content = self._strip_markdown(content)
                
                self.output_file.write_block(content)
                
                # Update statistics
                self.file_count += 1
                self.total_bytes += os.path.getsize(filepath)
        except Exception as e:
            self.output_file.write_line(f"[Error reading file: {str(e)}]")
    
    def _strip_markdown(self, content: str) -> str:
        """Strip basic markdown formatting from a string."""
        # Remove markdown headers
        content = re.sub(r'^#+\s*', '', content, flags=re.MULTILINE)
        # Remove bold: **text**
        content = re.sub(r'\*\*([^*]+)\*\*', r'\1', content)
        # Remove italic: *text*
        content = re.sub(r'\*([^*]+)\*', r'\1', content)
        # Remove underline: _text_
        content = re.sub(r'_([^_]+)_', r'\1', content)
        # Remove links: [text](url)
        content = re.sub(r'\[([^\]]+)\]\([^)]*\)', r'\1', content)
        # Remove code blocks (from ``` to ```)
        content = re.sub(r'```.*?```', '', content, flags=re.DOTALL)
        # Remove inline code: `text`
        content = re.sub(r'`([^`]+)`', r'\1', content)
        return content
    
    def _generate_summary(self) -> str:
        """Generate a summary of the repository analysis."""
        approx_tokens = self.total_bytes // 4  # Rough estimate
        return (
            "Repository Summary:\n"
            f"Files analyzed: {self.file_count}\n"
            f"Directories scanned: {self.dir_count}\n"
            f"Total size: {self.total_bytes} bytes\n"
            f"Estimated tokens: {approx_tokens}\n\n"
        )
    
    def process(self):
        """Process the source and generate the digest."""
        try:
            # Write Directory Structure section
            self.output_file.write_line("")
            self.output_file.write_line("# Directory Structure")
            self._process_directory(self.project_dir)
            
            # Write Files Content section
            self.output_file.write_line("")
            self.output_file.write_line("# Files Content")
            
            # Walk through all files in the project
            for root, _, files in os.walk(self.project_dir):
                for filename in files:
                    # Skip hidden files
                    if filename.startswith('.'):
                        continue
                    
                    filepath = os.path.join(root, filename)
                    rel_path = self._get_relative_path(filepath)
                    
                    # Skip the output file itself
                    if os.path.basename(filepath) == os.path.basename(self.output):
                        continue
                    
                    # Skip ignored files
                    if self._should_ignore(rel_path):
                        continue
                    
                    # Check if it's a supported file type
                    ext = os.path.splitext(filename)[1][1:].lower()
                    if ext in SUPPORTED_FILE_EXTENSIONS:
                        self._process_file_contents(filepath)
            
            # Generate summary
            summary = self._generate_summary()
            
            # Prepend summary to output file
            temp_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
            temp_file_path = temp_file.name
            
            with open(temp_file_path, 'w') as temp:
                temp.write(summary)
                with open(self.output_file.get_path(), 'r') as original:
                    temp.write(original.read())
            
            shutil.move(temp_file_path, self.output_file.get_path())
            
            print(f"Documentation has been generated at '{self.output_file.get_path()}'.", file=sys.stderr)
            
        finally:
            # Clean up temporary repository if one was created
            if self.temp_repo_path and os.path.exists(self.temp_repo_path):
                print(f"Cleaning up temporary repository at {self.temp_repo_path}", file=sys.stderr)
                shutil.rmtree(self.temp_repo_path)


def main():
    """Parse arguments and run the flattener."""
    parser = argparse.ArgumentParser(
        description="SourceCodeFlatener - Python Implementation"
    )
    
    parser.add_argument(
        "-s", "--source", required=True,
        help="Source (local directory or Git repository URL)"
    )
    parser.add_argument(
        "-o", "--output", required=True,
        help="Output file (base name; parts will have _partN appended if rotated)"
    )
    parser.add_argument(
        "-n", "--project-name",
        help="Project name (optional)"
    )
    parser.add_argument(
        "-f", "--max-file-size", type=int, default=DEFAULT_MAX_FILE_SIZE,
        help=f"Max file size for analysis in bytes (default: {DEFAULT_MAX_FILE_SIZE})"
    )
    parser.add_argument(
        "-O", "--max-output-size", type=int, default=DEFAULT_MAX_OUTPUT_SIZE,
        help=f"Max output file size in bytes before rotating (default: {DEFAULT_MAX_OUTPUT_SIZE})"
    )
    parser.add_argument(
        "-i", "--include-patterns", 
        help="Include patterns (comma-separated, e.g. '*.ps1,src/*')"
    )
    parser.add_argument(
        "-e", "--exclude-patterns",
        help=f"Exclude patterns (comma-separated, default: {','.join(DEFAULT_EXCLUDE_PATTERNS)})"
    )
    parser.add_argument(
        "-b", "--branch",
        help="Git branch (if cloning a repository)"
    )
    
    args = parser.parse_args()
    
    # Process comma-separated include/exclude patterns
    if args.include_patterns:
        args.include_patterns = [p.strip() for p in args.include_patterns.split(',')]
    
    if args.exclude_patterns:
        args.exclude_patterns = [p.strip() for p in args.exclude_patterns.split(',')]
    
    # Ensure output file has an extension (default to .md if none)
    if not os.path.splitext(args.output)[1]:
        args.output += ".md"
    
    # Run the flattener
    flattener = SourceCodeFlatener(args)
    flattener.process()


if __name__ == "__main__":
    main()

File: csharp/CodeDigest.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace SourceCodeFlatener
{
    class Program
    {
        // Default settings
        private const long DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024; // 10 MB
        private const long DEFAULT_MAX_OUTPUT_SIZE = 5 * 1024 * 1024; // 5 MB

        private static readonly HashSet<string> SUPPORTED_FILE_EXTENSIONS = new HashSet<string>
        {
            ".ps1", ".cs", ".sln", ".md", ".txt", ".json", ".xml", ".yaml", ".yml", ".py", ".js"
        };

        private static readonly HashSet<string> DEFAULT_EXCLUDE_PATTERNS = new HashSet<string>
        {
            "bin/*", "obj/*", "debug/*", "release/*", "node_modules/*",
            "*.dll", "*.exe", "*.pdb", "*.cache"
        };

        static async Task Main(string[] args)
        {
            var options = ParseCommandLine(args);
            if (options == null)
            {
                PrintUsage();
                return;
            }

            var flattener = new SourceCodeFlatener(options);
            await flattener.ProcessAsync();
        }

        static void PrintUsage()
        {
            Console.WriteLine("SourceCodeFlatener - C# Implementation");
            Console.WriteLine("Usage: dotnet CodeDigest.dll -s <source> -o <output_file> [options]");
            Console.WriteLine();
            Console.WriteLine("Parameters:");
            Console.WriteLine("  -s, --source         Source (local directory or Git repository URL)");
            Console.WriteLine("  -o, --output         Output file (base name; parts will have _partN appended if rotated)");
            Console.WriteLine("  -n, --project-name   Project name (optional)");
            Console.WriteLine($"  -f, --max-file-size  Max file size for analysis in bytes (default: {DEFAULT_MAX_FILE_SIZE})");
            Console.WriteLine($"  -O, --max-output-size Max output file size in bytes before rotating (default: {DEFAULT_MAX_OUTPUT_SIZE})");
            Console.WriteLine("  -i, --include        Include patterns (comma-separated, e.g. \"*.ps1,src/*\")");
            Console.WriteLine("  -e, --exclude        Exclude patterns (comma-separated)");
            Console.WriteLine("  -b, --branch         Git branch (if cloning a repository)");
        }

        static Options ParseCommandLine(string[] args)
        {
            var options = new Options
            {
                MaxFileSize = DEFAULT_MAX_FILE_SIZE,
                MaxOutputSize = DEFAULT_MAX_OUTPUT_SIZE,
                ExcludePatterns = new HashSet<string>(DEFAULT_EXCLUDE_PATTERNS)
            };

            for (int i = 0; i < args.Length; i++)
            {
                string arg = args[i].ToLower();
                string nextArg = (i + 1 < args.Length) ? args[i + 1] : null;

                if (string.IsNullOrEmpty(nextArg) || nextArg.StartsWith("-"))
                    nextArg = null;

                switch (arg)
                {
                    case "-s":
                    case "--source":
                        if (nextArg != null)
                        {
                            options.Source = nextArg;
                            i++;
                        }
                        break;
                    case "-o":
                    case "--output":
                        if (nextArg != null)
                        {
                            options.Output = nextArg;
                            i++;
                        }
                        break;
                    case "-n":
                    case "--project-name":
                        if (nextArg != null)
                        {
                            options.ProjectName = nextArg;
                            i++;
                        }
                        break;
                    case "-f":
                    case "--max-file-size":
                        if (nextArg != null && long.TryParse(nextArg, out long maxFileSize))
                        {
                            options.MaxFileSize = maxFileSize;
                            i++;
                        }
                        break;
                    case "-o":
                    case "--max-output-size":
                        if (nextArg != null && long.TryParse(nextArg, out long maxOutputSize))
                        {
                            options.MaxOutputSize = maxOutputSize;
                            i++;
                        }
                        break;
                    case "-i":
                    case "--include":
                        if (nextArg != null)
                        {
                            options.IncludePatterns = new HashSet<string>(
                                nextArg.Split(',').Select(p => p.Trim()).Where(p => !string.IsNullOrWhiteSpace(p))
                            );
                            i++;
                        }
                        break;
                    case "-e":
                    case "--exclude":
                        if (nextArg != null)
                        {
                            options.ExcludePatterns = new HashSet<string>(
                                nextArg.Split(',').Select(p => p.Trim()).Where(p => !string.IsNullOrWhiteSpace(p))
                            );
                            i++;
                        }
                        break;
                    case "-b":
                    case "--branch":
                        if (nextArg != null)
                        {
                            options.Branch = nextArg;
                            i++;
                        }
                        break;
                }
            }

            // Validate required parameters
            if (string.IsNullOrEmpty(options.Source) || string.IsNullOrEmpty(options.Output))
                return null;

            // Ensure output file has an extension (default to .md if none)
            if (!Path.HasExtension(options.Output))
                options.Output += ".md";

            return options;
        }
    }

    class Options
    {
        public string Source { get; set; }
        public string Output { get; set; }
        public string ProjectName { get; set; }
        public long MaxFileSize { get; set; }
        public long MaxOutputSize { get; set; }
        public HashSet<string> IncludePatterns { get; set; } = new HashSet<string>();
        public HashSet<string> ExcludePatterns { get; set; } = new HashSet<string>();
        public string Branch { get; set; }
    }

    class OutputFile
    {
        private readonly string _basePath;
        private readonly long _maxSize;
        private readonly string _projectName;
        private int _part = 1;
        private string _currentPath;
        private readonly string _baseName;
        private readonly string _extension;
        private readonly string _directory;

        public OutputFile(string outputPath, long maxSize, string projectName)
        {
            _basePath = Path.GetFullPath(outputPath);
            _maxSize = maxSize;
            _projectName = projectName ?? "Unnamed Project";

            _directory = Path.GetDirectoryName(_basePath);
            _baseName = Path.GetFileNameWithoutExtension(_basePath);
            _extension = Path.GetExtension(_basePath);
            _currentPath = _basePath;

            // Create directory if it doesn't exist
            Directory.CreateDirectory(_directory);

            // Write initial header
            using (var writer = new StreamWriter(_currentPath, false, Encoding.UTF8))
            {
                writer.WriteLine($"# Project Digest: {_projectName}");
                writer.WriteLine($"Generated on: {DateTime.Now}");
                writer.WriteLine();
            }
        }

        private void CheckAndRotate()
        {
            if (!File.Exists(_currentPath))
                return;

            var fileInfo = new FileInfo(_currentPath);
            if (fileInfo.Length > _maxSize)
            {
                _part++;
                _currentPath = Path.Combine(_directory, $"{_baseName}_part{_part}{_extension}");

                // Write header to new part
                using (var writer = new StreamWriter(_currentPath, false, Encoding.UTF8))
                {
                    writer.WriteLine($"# Project Digest Continued: {_projectName}");
                    writer.WriteLine($"Generated on: {DateTime.Now}");
                    writer.WriteLine();
                }
            }
        }

        public void WriteLine(string line)
        {
            using (var writer = new StreamWriter(_currentPath, true, Encoding.UTF8))
            {
                writer.WriteLine(line);
            }
            CheckAndRotate();
        }

        public void WriteBlock(string block)
        {
            foreach (var line in block.Split('\n'))
            {
                WriteLine(line.TrimEnd('\r'));
            }
        }

        public string GetPath() => _currentPath;
    }

    class SourceCodeFlatener
    {
        private readonly Options _options;
        private readonly string _projectDir;
        private readonly OutputFile _outputFile;
        private readonly string _tempRepoPath;
        private int _fileCount = 0;
        private int _dirCount = 0;
        private long _totalBytes = 0;

        public SourceCodeFlatener(Options options)
        {
            _options = options;

            // Determine project directory
            if (options.Source.StartsWith("http://") || options.Source.StartsWith("https://"))
                _projectDir = _tempRepoPath = CloneRepository();
            else
            {
                if (!Directory.Exists(options.Source))
                    throw new DirectoryNotFoundException($"Directory '{options.Source}' does not exist.");
                _projectDir = Path.GetFullPath(options.Source);
            }

            // Initialize output file
            _outputFile = new OutputFile(
                options.Output,
                options.MaxOutputSize,
                options.ProjectName
            );

            // Update output file with source info
            _outputFile.WriteLine($"Source: {options.Source}");
            _outputFile.WriteLine($"Project Directory: {_projectDir}");
            _outputFile.WriteLine("");
        }

        private string CloneRepository()
        {
            Console.Error.WriteLine($"Source appears to be a URL: {_options.Source}");
            string tempPath = Path.Combine(Path.GetTempPath(), $"repo_{Guid.NewGuid()}");
            Directory.CreateDirectory(tempPath);
            
            Console.Error.WriteLine($"Cloning repository {_options.Source} into {tempPath} ...");

            string args = $"clone ";
            if (!string.IsNullOrEmpty(_options.Branch))
                args += $"--branch {_options.Branch} ";
            else
                args += "--depth=1 ";

            args += $"{_options.Source} {tempPath}";

            var processInfo = new ProcessStartInfo
            {
                FileName = "git",
                Arguments = args,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            try
            {
                using (var process = Process.Start(processInfo))
                {
                    process.WaitForExit();
                    if (process.ExitCode != 0)
                    {
                        string error = process.StandardError.ReadToEnd();
                        throw new Exception($"Error cloning repository: {error}");
                    }
                }
            }
            catch (Exception ex)
            {
                if (Directory.Exists(tempPath))
                    Directory.Delete(tempPath, true);
                throw new Exception($"Failed to clone repository: {ex.Message}");
            }

            return tempPath;
        }

        private bool ShouldIgnore(string relPath)
        {
            // If include patterns are specified and path doesn't match any, ignore it
            if (_options.IncludePatterns.Count > 0)
            {
                bool matchFound = false;
                foreach (var pattern in _options.IncludePatterns)
                {
                    if (GlobMatch(relPath, pattern))
                    {
                        matchFound = true;
                        break;
                    }
                }
                if (!matchFound)
                    return true;
            }

            // Check if path matches any exclude pattern
            foreach (var pattern in _options.ExcludePatterns)
            {
                if (GlobMatch(relPath, pattern))
                    return true;
            }

            return false;
        }

        private bool GlobMatch(string path, string pattern)
        {
            // Convert glob pattern to regex
            string regex = "^" + Regex.Escape(pattern).Replace("\\*", ".*") + "$";
            return Regex.IsMatch(path, regex);
        }

        private string GetRelativePath(string path) =>
            Path.GetRelativePath(_projectDir, path);

        private async Task ProcessDirectoryAsync(string directory, string indent = "")
        {
            string relPath = GetRelativePath(directory);

            if (relPath == ".")
                _outputFile.WriteLine("[DIR] .");
            else
            {
                string dirName = Path.GetFileName(directory);
                _outputFile.WriteLine($"{indent}[DIR] {dirName}");
            }

            _dirCount++;

            // Process directory contents
            foreach (var item in Directory.GetFileSystemEntries(directory))
            {
                // Skip hidden files/directories
                string itemName = Path.GetFileName(item);
                if (itemName.StartsWith("."))
                    continue;

                string itemRel = GetRelativePath(item);

                // Skip the output file itself
                if (Path.GetFileName(item) == Path.GetFileName(_options.Output))
                    continue;

                // Skip ignored files/directories
                if (ShouldIgnore(itemRel))
                    continue;

                if (Directory.Exists(item))
                    await ProcessDirectoryAsync(item, $"  {indent}");
                else
                {
                    // Check if it's a supported file type
                    string ext = Path.GetExtension(item).ToLowerInvariant();
                    if (SUPPORTED_FILE_EXTENSIONS.Contains(ext))
                        _outputFile.WriteLine($"{indent}  [FILE] {itemName}");
                }
            }
        }

        private async Task ProcessFileContentsAsync(string filepath)
        {
            string relPath = GetRelativePath(filepath);
            string ext = Path.GetExtension(filepath).ToLowerInvariant();

            if (!SUPPORTED_FILE_EXTENSIONS.Contains(ext))
                return;

            // Skip files that are too large
            var fileInfo = new FileInfo(filepath);
            if (fileInfo.Length > _options.MaxFileSize)
            {
                _outputFile.WriteLine("");
                _outputFile.WriteLine($"## {relPath}");
                _outputFile.WriteLine("[File too large to process]");
                return;
            }

            _outputFile.WriteLine("");
            _outputFile.WriteLine($"## {relPath}");

            try
            {
                string content = await File.ReadAllTextAsync(filepath);

                // For markdown files, we might want to strip formatting
                if (ext == ".md")
                    content = StripMarkdown(content);

                _outputFile.WriteBlock(content);

                // Update statistics
                _fileCount++;
                _totalBytes += fileInfo.Length;
            }
            catch (Exception ex)
            {
                _outputFile.WriteLine($"[Error reading file: {ex.Message}]");
            }
        }

        private string StripMarkdown(string content)
        {
            // Remove markdown headers
            content = Regex.Replace(content, @"^#+\s*", "", RegexOptions.Multiline);
            // Remove bold: **text**
            content = Regex.Replace(content, @"\*\*([^*]+)\*\*", "$1");
            // Remove italic: *text*
            content = Regex.Replace(content, @"\*([^*]+)\*", "$1");
            // Remove underline: _text_
            content = Regex.Replace(content, @"_([^_]+)_", "$1");
            // Remove links: [text](url)
            content = Regex.Replace(content, @"\[([^\]]+)\]\([^)]*\)", "$1");
            // Remove code blocks (from ``` to ```)
            content = Regex.Replace(content, @"```.*?```", "", RegexOptions.Singleline);
            // Remove inline code: `text`
            content = Regex.Replace(content, @"`([^`]+)`", "$1");
            return content;
        }

        private string GenerateSummary()
        {
            long approxTokens = _totalBytes / 4; // Rough estimate
            return
                "Repository Summary:\n" +
                $"Files analyzed: {_fileCount}\n" +
                $"Directories scanned: {_dirCount}\n" +
                $"Total size: {_totalBytes} bytes\n" +
                $"Estimated tokens: {approxTokens}\n\n";
        }

        public async Task ProcessAsync()
        {
            try
            {
                // Write Directory Structure section
                _outputFile.WriteLine("");
                _outputFile.WriteLine("# Directory Structure");
                await ProcessDirectoryAsync(_projectDir);

                // Write Files Content section
                _outputFile.WriteLine("");
                _outputFile.WriteLine("# Files Content");

                // Walk through all files in the project
                foreach (var filepath in Directory.GetFiles(_projectDir, "*", SearchOption.AllDirectories))
                {
                    // Skip hidden files
                    string filename = Path.GetFileName(filepath);
                    if (filename.StartsWith("."))
                        continue;

                    string relPath = GetRelativePath(filepath);

                    // Skip the output file itself
                    if (Path.GetFileName(filepath) == Path.GetFileName(_options.Output))
                        continue;

                    // Skip ignored files
                    if (ShouldIgnore(relPath))
                        continue;

                    // Check if it's a supported file type
                    string ext = Path.GetExtension(filepath).ToLowerInvariant();
                    if (SUPPORTED_FILE_EXTENSIONS.Contains(ext))
                        await ProcessFileContentsAsync(filepath);
                }

                // Generate summary
                string summary = GenerateSummary();

                // Prepend summary to output file
                string tempFilePath = Path.GetTempFileName();
                try
                {
                    using (var writer = new StreamWriter(tempFilePath, false, Encoding.UTF8))
                    {
                        writer.Write(summary);
                        writer.Write(await File.ReadAllTextAsync(_outputFile.GetPath()));
                    }

                    File.Move(tempFilePath, _outputFile.GetPath(), true);
                }
                finally
                {
                    if (File.Exists(tempFilePath))
                        File.Delete(tempFilePath);
                }

                Console.Error.WriteLine($"Documentation has been generated at '{_outputFile.GetPath()}'.");
            }
            finally
            {
                // Clean up temporary repository if one was created
                if (!string.IsNullOrEmpty(_tempRepoPath) && Directory.Exists(_tempRepoPath))
                {
                    Console.Error.WriteLine($"Cleaning up temporary repository at {_tempRepoPath}");
                    Directory.Delete(_tempRepoPath, true);
                }
            }
        }
    }
}
